lib/ae_mdw/mnesia.ex

```diff
@@ -0,0 +1,139 @@
+  def last_key(tab) do
+    case :mnesia.dirty_last(tab) do
+      @end_token -> :none
+      last_key -> {:ok, last_key}
+    end
+  end
+
+  @spec first_key(table()) :: {:ok, key()} | :none
+  def first_key(tab) do
+    case :mnesia.dirty_first(tab) do
+      @end_token -> :none
+      last_key -> {:ok, last_key}
+    end
+  end
+
+  @spec fetch(table(), key()) :: {:ok, record()} | :not_found
+  def fetch(tab, key) do
+    case :mnesia.dirty_read(tab, key) do
+      [record] -> {:ok, record}
+      [] -> :not_found
+    end
+  end
+
+  @spec fetch!(table(), key()) :: record()
+  def fetch!(tab, key) do
+    {:ok, record} = fetch(tab, key)
+
+    record
+  end
+
+  @spec fetch_keys(table(), direction(), cursor(), non_neg_integer()) :: {[record()], cursor()}
+  def fetch_keys(tab, :forward, nil, limit) do
+    {:atomic, {keys, cursor}} =
+      :mnesia.transaction(fn ->
+        case :mnesia.first(tab) do
+          @end_token -> {[], nil}
+          first_key -> fetch_forward_keys(tab, first_key, limit)
+        end
+      end)
+
+    {keys, cursor}
+  end
+
+  def fetch_keys(tab, :forward, first_key, limit) do
+    {:atomic, {keys, cursor}} =
+      :mnesia.transaction(fn ->
+        case :mnesia.read(tab, first_key) do
+          @end_token -> {[], nil}
+          _first_record -> fetch_forward_keys(tab, first_key, limit)
+        end
+      end)
+
+    {keys, cursor}
+  end
+
+  def fetch_keys(tab, :backward, nil, limit) do
+    {:atomic, {keys, cursor}} =
+      :mnesia.transaction(fn ->
+        case :mnesia.last(tab) do
+          @end_token -> {[], nil}
+          last_key -> fetch_backward_keys(tab, last_key, limit)
+        end
+      end)
+
+    {keys, cursor}
+  end
+
+  def fetch_keys(tab, :backward, last_key, limit) do
+    {:atomic, {keys, cursor}} =
+      :mnesia.transaction(fn ->
+        case :mnesia.read(tab, last_key) do
+          @end_token -> {[], nil}
+          _last_record -> fetch_backward_keys(tab, last_key, limit)
+        end
+      end)
+
+    {keys, cursor}
+  end
+
+  defp fetch_forward_keys(tab, first_key, limit) do
+    keys =
+      Stream.unfold({limit, first_key}, fn
+        {0, _current_key} ->
+          nil
+
+        {n, current_key} ->
+          case :mnesia.next(tab, current_key) do
+            @end_token -> nil
+            next_key -> {next_key, {n - 1, next_key}}
+          end
+      end)
+
+    case Enum.split(keys, limit - 1) do
+      {keys, [next_key]} -> {[first_key | keys], next_key}
+      {keys, []} -> {keys, nil}
+    end
+  end
+
+  defp fetch_backward_keys(tab, last_key, limit) do
+    keys =
+      Stream.unfold({limit, last_key}, fn
+        {0, _current_key} ->
+          nil
+
+        {n, current_key} ->
+          case :mnesia.prev(tab, current_key) do
+            @end_token -> nil
+            prev_key -> {prev_key, {n - 1, prev_key}}
+          end
+      end)
+
+    case Enum.split(keys, limit - 1) do
+      {keys, [prev_key]} -> {[last_key | keys], prev_key}
```

Comment:
considering limit = 10, keys has 9 elements after `Enum.split`

Reply:
`keys` will have 9 elements, while `[last_key | key]` will have 10, as requested, so we're good